###SQLite

SQLite is an in-process library that implements a self-contained, serverless, zero-configuration, transactional SQL database engine. The code for SQLite is in the public domain and is thus free for use for any purpose, commercial or private. SQLite is the most widely deployed database in the world with more applications than we can count, including several high-profile projects.

SQLite is an embedded SQL database engine. Unlike most other SQL databases, SQLite does not have a separate server process. SQLite reads and writes directly to ordinary disk files. A complete SQL database with multiple tables, indices, triggers, and views, is contained in a single disk file. The database file format is cross-platform - you can freely copy a database between 32-bit and 64-bit systems or between big-endian and little-endian architectures. These features make SQLite a popular choice as an Application File Format. Think of SQLite not as a replacement for Oracle but as a replacement for fopen()

Because SQLite uses a very generic type system, we don't get the strict data types that we would in most other databases (such as MySQL and PostgreSQL); therefore, all of our data is going to be stored as text. In other databases, we'd store dates as a date or datetime field but in SQLite it's text.

http://www.thegeekstuff.com/2012/09/sqlite-command-examples/

In [5]:
import pandas as pd
import sqlite3

# Create a SQL connection to our SQLite database
#con = sqlite3.connect(r'C:\Users\Sterling7\Dropbox\dave45678\Python for Data Analytics\sqlite3\Chinook_Sqlite.sqlite')
con = sqlite3.connect(r'sqlite3\movielens')
cur = con.cursor()

# the result of a "cursor.execute" can be iterated over by row
for row in cur.execute('SELECT * FROM Movies;'):
    print(row)



(1, 'Toy Story (1995)', '1995-01-01')
(2, 'GoldenEye (1995)', '1995-01-01')
(3, 'Four Rooms (1995)', '1995-01-01')
(4, 'Get Shorty (1995)', '1995-01-01')
(5, 'Copycat (1995)', '1995-01-01')
(6, 'Shanghai Triad (Yao a yao yao dao waipo qiao) (1995)', '1995-01-01')
(7, 'Twelve Monkeys (1995)', '1995-01-01')
(8, 'Babe (1995)', '1995-01-01')
(9, 'Dead Man Walking (1995)', '1995-01-01')
(10, 'Richard III (1995)', '1996-01-22')
(11, 'Seven (Se7en) (1995)', '1995-01-01')
(12, 'Usual Suspects, The (1995)', '1995-08-14')
(13, 'Mighty Aphrodite (1995)', '1995-10-30')
(14, 'Postino, Il (1994)', '1994-01-01')
(15, "Mr. Holland's Opus (1995)", '1996-01-29')
(16, 'French Twist (Gazon maudit) (1995)', '1995-01-01')
(17, 'From Dusk Till Dawn (1996)', '1996-02-05')
(18, 'White Balloon, The (1995)', '1995-01-01')
(19, "Antonia's Line (1995)", '1995-01-01')
(20, 'Angels and Insects (1995)', '1995-01-01')
(21, 'Muppet Treasure Island (1996)', '1996-02-16')
(22, 'Braveheart (1995)', '1996-02-16')
(23, 'Tax

In [6]:
df_movies = pd.read_sql_query("SELECT * from Movies;", con)

# verify that result of SQL query is stored in the dataframe
print(df_movies.head())

#Be sure to close the connection.
con.close()

   id              title release_date
0   1   Toy Story (1995)   1995-01-01
1   2   GoldenEye (1995)   1995-01-01
2   3  Four Rooms (1995)   1995-01-01
3   4  Get Shorty (1995)   1995-01-01
4   5     Copycat (1995)   1995-01-01


In [7]:
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1682 entries, 0 to 1681
Data columns (total 3 columns):
id              1682 non-null int64
title           1682 non-null object
release_date    1681 non-null object
dtypes: int64(1), object(2)
memory usage: 52.6+ KB


The output tells a few things about our DataFrame.
Each row was assigned an index of 0 to N-1, where N is the number of rows in the DataFrame. pandas will do this by default if an index is not specified. Don't worry, this can be changed later.
There are 1,682 rows (every row must have an index).
Our dataset has five total columns, one of which isn't populated at all (video_release_date) and two that are missing some values (release_date and imdb_url).
The last line displays the datatypes of each column, but not necessarily in the corresponding order to the listed columns.
You should use the dtypes method to get the datatype for each column.

In [8]:
df_movies.dtypes

id               int64
title           object
release_date    object
dtype: object

Benchmarking our results
http://sebastianraschka.com/Articles/sqlite3_database.html#benchmarks

Where clause examples

towed[towed["state"] == "TX"] # all columns and records where the car was from TX

towed[(towed["state"] == "TX") & (towed["make"] == "KIA")] # made by KIA AND from TX

towed[(towed["state"] == "MA") | (towed["make"] == "JAGU")] # made by Jaguar OR from MA

towed[towed["make"] == "KIA"].sort("color") # made by KIA, ordered by color (A to Z)